keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,750 @@
+* while keeping needed expressivity to express Kubernetes-like APIs including powerful value validations.
+
+Schemas created through code generation (like [crd-gen](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/crd) and [openapi-gen](https://github.com/kubernetes/kube-openapi/tree/master/cmd/openapi-gen)) which recurses over API Golang types essentially produce structural schemas naturally. Developer can enrich those structural schemas with nearly arbitrary value validation logic, usually using in-code `+kubebuilder`-like tags.
+
+The restriction is an API change of the CustomResourceDefinition kind in the `v1beta1` `apiextensions.k8s.io` API group, but in the sense that
+* CRDs created via the `v1beta1` endpoints are not restricted such that existing CRDs keep working **as before**,
+* but CRDs created or updated via the `v1beta1` endpoints have to conform to the new restrictions **if they want to make use of new features**.
+
+In the `v1` CustomResourceDefinition API we will enforce structural schemas on creation and do ratcheting validation on update (allow updates with non-structural schemas if the original schema was non-structural already).
+
+To communicate to the developer that a CRD does not follow the structural schema restriction and will therefore not benefit from future developments, we add a `NonStructuralSchema` condition to the CRD.
+
+## Motivation
+
+* OpenAPI has the goal of describing every API possible. We have the goal of presenting a consistent API to users of Kubernetes. OpenAPI therefore makes a fantastic output format able to describe anything, but a poor input format, because there's not "one correct way" to do a given thing. It is easy to accidentally (or intentionally) describe things in OpenAPI v3 that are not similar to other Kubernetes APIs and far outside of the [Kuberntes API conversions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md).
+* OpenAPI v3 is a very open and wide spec that can define crazy things using logical operators (`not(anyOf(oneof ...))`), leading to complicated semantics, making advanced CRD features difficult or even impossible to implement.
+
+### Goals
+
+* Restrict CRD validation OpenAPI v3 schemas to give one canonical way to express objects we intend to support as Kubernetes-like types.
+* Simplify the semantics of CRD validation OpenAPI v3 schemas to make implementation of advanced feature like pruning, defaulting, server-side apply and read-only validation considerably less difficult.
+* Support embedding of existing API types into a CRD via RawExtensions or via Golang type reuse (e.g. of PodSpecs).
+* Allow a rich OpenAPI v3 validation schema for CustomResource validation, but allow us to programmatically derive the restricted spec for all other purposes.
+* Perspective goal: have a canonical mapping from a restricted schema to Golang types and Protobuf v3.
```

Comment:
s/Perspective/Prospective/?

Reply:
if that's a known term, yes. Some goal to align with future direction.